// nickjs/lib/es5/Nick.js
"use strict";var _getIterator2 = require("babel-runtime/core-js/get-iterator");var _getIterator3 = _interopRequireDefault(_getIterator2);var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require("babel-runtime/helpers/createClass");var _createClass3 = _interopRequireDefault(_createClass2);var _stringify = require("babel-runtime/core-js/json/stringify");var _stringify2 = _interopRequireDefault(_stringify);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}var Promise = require("bluebird");
var _ = require("lodash");
var once = require("once");
var toBoolean = require("to-boolean");
var Tab = require("./Tab");

Promise.onPossiblyUnhandledRejection(function (e, promise) {

	console.log("> Unhandled Promise Rejection:");
	console.log(e);
	console.log((0, _stringify2.default)(e, undefined, 2));
	throw e;
});var

Nick = function () {

	function Nick() {var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};(0, _classCallCheck3.default)(this, Nick);

		if (typeof process !== "undefined" && _.isObject(process) && _.isObject(process.versions) && typeof process.versions.node === 'string') {
			var BrowserDriver = require('./chrome/BrowserDriver');
			var environment = process.env;
		} else if (typeof phantom !== "undefined" && _.isObject(phantom) && typeof phantom.casperPath === 'string') {
			var BrowserDriver = require('./casper/BrowserDriver');
			var environment = require("system").env;
		} else {
			if (_.isObject(phantom)) {
				throw new Error("Cannot initialize NickJS because it seems you're running PhantomJS without CasperJS. NickJS scripts can be run by either NodeJS or CasperJS, but not PhantomJS alone.");
			} else {
				throw new Error("Cannot initialize NickJS: could not determine the environment. Is it NodeJS or CasperJS? Where are we? NickJS scripts can be run by either NodeJS or CasperJS.");
			}
		}


		if (!_.isPlainObject(options)) {
			throw new TypeError('options must be of type plain object');
		}


		if (_.has(options, 'debug')) {
			if (typeof options.debug !== 'boolean') {
				throw new TypeError('debug option must be of type boolean');
			}
		} else {
			options.debug = false;
		}


		if (_.has(options, 'headless')) {
			if (typeof options.headless !== 'boolean') {
				throw new TypeError('headless option must be of type boolean');
			}
		} else {
			options.headless = true;
		}


		if (_.has(options, 'loadImages')) {
			if (typeof options.loadImages !== 'boolean') {
				throw new TypeError('loadImages option must be of type boolean');
			}
		} else if (environment.NICKJS_LOAD_IMAGES) {
			options.loadImages = toBoolean(environment.NICKJS_LOAD_IMAGES || false);
		} else {


		}


		if (_.has(options, 'httpProxy')) {
			if (typeof options.httpProxy !== 'string') {
				throw new TypeError('httpProxy option must be of type string');
			}
		} else {
			options.httpProxy = environment.NICKJS_PROXY || environment.HTTP_PROXY || environment.http_proxy;
		}

		if (options.httpProxy) {
			var urlCheck = options.httpProxy.trim().toLowerCase();
			if (urlCheck.length) {
				if (urlCheck.indexOf("http://") < 0 && urlCheck.indexOf("https://") < 0) {
					options.httpProxy = "http://" + options.httpProxy;
				}
			} else {
				options.httpProxy = null;
			}
		}


		if (_.has(options, 'userAgent')) {
			if (typeof options.userAgent !== 'string') {
				throw new TypeError('userAgent option must be of type string');
			}
		} else {
			options.userAgent = 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_13_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.146 Safari/537.36';
		}


		if (_.has(options, 'timeout')) {
			if (typeof options.timeout !== 'number' || options.timeout < 0) {
				throw new TypeError('timeout option must be a positive number');
			}
		} else {
			options.timeout = 10000;
		}


		if (_.has(options, 'width')) {
			if (typeof options.width !== 'number' || options.width < 0) {
				throw new TypeError('width option must be a positive number');
			}
		} else {
			options.width = 1280;
		}


		if (_.has(options, 'height')) {
			if (typeof options.height !== 'number' || options.height < 0) {
				throw new TypeError('height option must be a positive number');
			}
		} else {
			options.height = 800;
		}


		if (_.has(options, 'printNavigation')) {
			if (typeof options.printNavigation !== 'boolean') {
				throw new TypeError('printNavigation option must be of type boolean');
			}
		} else {
			options.printNavigation = true;
		}


		if (_.has(options, 'printPageErrors')) {
			if (typeof options.printPageErrors !== 'boolean') {
				throw new TypeError('printPageErrors option must be of type boolean');
			}
		} else {
			options.printPageErrors = true;
		}


		if (_.has(options, 'printResourceErrors')) {
			if (typeof options.printResourceErrors !== 'boolean') {
				throw new TypeError('printResourceErrors option must be of type boolean');
			}
		} else {
			options.printResourceErrors = true;
		}


		if (_.has(options, 'printAborts')) {
			if (typeof options.printAborts !== 'boolean') {
				throw new TypeError('printAborts option must be of type boolean');
			}
		} else {
			options.printAborts = true;
		}


		var whitelist = [];
		if (_.has(options, 'whitelist')) {
			if (_.isArray(options.whitelist)) {var _iteratorNormalCompletion = true;var _didIteratorError = false;var _iteratorError = undefined;try {
					for (var _iterator = (0, _getIterator3.default)(options.whitelist), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {var white = _step.value;
						if (white instanceof RegExp) {
							whitelist.push(white);
						} else if (typeof white === 'string') {
							whitelist.push(white.toLowerCase());
						} else {
							throw new TypeError('whitelist option must be an array of strings or regexes');
						}
					}} catch (err) {_didIteratorError = true;_iteratorError = err;} finally {try {if (!_iteratorNormalCompletion && _iterator.return) {_iterator.return();}} finally {if (_didIteratorError) {throw _iteratorError;}}}
			} else {
				throw new TypeError('whitelist option must be an array of strings or regexes');
			}
		}
		options.whitelist = whitelist;


		var blacklist = [];
		if (_.has(options, 'blacklist')) {
			if (_.isArray(options.blacklist)) {var _iteratorNormalCompletion2 = true;var _didIteratorError2 = false;var _iteratorError2 = undefined;try {
					for (var _iterator2 = (0, _getIterator3.default)(options.blacklist), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {var black = _step2.value;
						if (black instanceof RegExp) {
							blacklist.push(black);
						} else if (typeof black === 'string') {
							blacklist.push(black.toLowerCase());
						} else {
							throw new TypeError('blacklist option must be an array of strings or regexes');
						}
					}} catch (err) {_didIteratorError2 = true;_iteratorError2 = err;} finally {try {if (!_iteratorNormalCompletion2 && _iterator2.return) {_iterator2.return();}} finally {if (_didIteratorError2) {throw _iteratorError2;}}}
			} else {
				throw new TypeError('blacklist option must be an array of strings or regexes');
			}
		}
		options.blacklist = blacklist;

		this._options = _.cloneDeep(options);

		this._initialized = false;
		this._initializing = false;
		this._tabIdCounter = 0;

		this._tabs = {};



		this._browserDriver = new BrowserDriver(this);
	}(0, _createClass3.default)(Nick, [{ key: "exit", value: function exit()







		{var code = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 0;
			this._browserDriver.exit(code);
		} }, { key: "initialize", value: function initialize()






		{var _this = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			var promise = new Promise(function (fulfill, reject) {
				if (_this._initialized) {
					fulfill(null);
				} else {
					if (_this._initializing) {
						var checkForInitialization = function checkForInitialization() {
							setTimeout(function () {
								if (_this._initializing) {
									checkForInitialization();
								} else {
									if (_this._initialized) {
										fulfill(null);
									} else {
										reject('browser initialization failed');
									}
								}
							}, 250);
						};
						checkForInitialization();
					} else {
						_this._initializing = true;
						_this._browserDriver._initialize(once(function (err) {
							_this._initializing = false;
							if (err) {
								reject(err);
							} else {
								_this._initialized = true;
								fulfill(null);
							}
						}));
					}
				}
			});
			return promise.asCallback(callback);
		} }, { key: "newTab", value: function newTab()

		{var _this2 = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			return this.initialize().then(function () {
				return new Promise(function (fulfill, reject) {
					++_this2._tabIdCounter;
					_this2._browserDriver._newTabDriver(_this2._tabIdCounter, function (err, tabDriver) {
						if (err) {
							reject(err);
						} else {
							var t = new Tab(_this2, tabDriver);
							_this2._tabs["" + _this2._tabIdCounter] = t;
							fulfill(t);
						}
					});
				});
			}).asCallback(callback);
		} }, { key: "unrefTabById", value: function unrefTabById(

		id) {
			delete this._tabs["" + id];
		} }, { key: "getAllCookies", value: function getAllCookies()

		{var _this3 = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			return this.initialize().then(function () {return Promise.fromCallback(function (callback) {return _this3._browserDriver._getAllCookies(callback);});}).asCallback(callback);
		} }, { key: "deleteAllCookies", value: function deleteAllCookies()

		{var _this4 = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			return this.initialize().then(function () {return Promise.fromCallback(function (callback) {return _this4._browserDriver._deleteAllCookies(callback);});}).asCallback(callback);
		} }, { key: "deleteCookie", value: function deleteCookie(

		name, domain) {var _this5 = this;var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
			if (typeof name !== 'string') {
				throw new TypeError('deleteCookie: name parameter must be of type string');
			}
			if (typeof domain !== 'string') {
				throw new TypeError('deleteCookie: domain parameter must be of type string');
			}
			return this.initialize().then(function () {return Promise.fromCallback(function (callback) {return _this5._browserDriver._deleteCookie(name, domain, callback);});}).asCallback(callback);
		} }, { key: "setCookie", value: function setCookie(

		cookie) {var _this6 = this;var callback = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
			if (!_.isPlainObject(cookie)) {
				throw new TypeError('setCookie: cookie parameter must be of type plain object');
			}
			if (_.has(cookie, 'domain')) {
				if (typeof cookie.domain !== 'string') {
					throw new TypeError('setCookie: cookie name option must be of type string');
				}
			}
			if (typeof cookie.name !== 'string') {
				throw new TypeError('setCookie: cookie name option must be of type string');
			}
			if (typeof cookie.value !== 'string') {
				throw new TypeError('setCookie: cookie value option must be of type string');
			}
			if (_.has(cookie, 'httponly')) {
				cookie.httpOnly = cookie.httponly;
			}
			if (_.has(cookie, 'httpOnly')) {
				if (typeof cookie.httpOnly !== 'boolean') {
					throw new TypeError('setCookie: cookie httpOnly option must be of type boolean');
				}
			}
			if (_.has(cookie, 'secure')) {
				if (typeof cookie.secure !== 'boolean') {
					throw new TypeError('setCookie: cookie secure option must be of type boolean');
				}
			}
			return this.initialize().then(function () {return Promise.fromCallback(function (callback) {return _this6._browserDriver._setCookie(cookie, callback);});}).asCallback(callback);
		} }, { key: "driver", get: function get() {return this._browserDriver;} }, { key: "browserDriver", get: function get() {return this._browserDriver;} }, { key: "options", get: function get() {return this._options;} }, { key: "tabs", get: function get() {return this._tabs;} }]);return Nick;}();



module.exports = Nick;
