// nickjs/lib/es5/Tab.js
"use strict";var _getIterator2 = require("babel-runtime/core-js/get-iterator");var _getIterator3 = _interopRequireDefault(_getIterator2);var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require("babel-runtime/helpers/createClass");var _createClass3 = _interopRequireDefault(_createClass2);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}var Promise = require("bluebird");
var _ = require("lodash");var

Tab = function () {

	function Tab(nick, tabDriver) {(0, _classCallCheck3.default)(this, Tab);
		this._nick = nick;
		this._tabDriver = tabDriver;
		this._actionInProgress = false;

		this._tabDriver._onConfirm = function (msg) {
			return true;
		};
		this._tabDriver._onPrompt = function (msg) {
			return "";
		};
	}(0, _createClass3.default)(Tab, [{ key: "_callToTabDriver", value: function _callToTabDriver(
























		action, callback) {var _this = this;var multiArgs = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : false;
			if (this._tabDriver.crashed) {
				throw new Error('this tab has crashed, no other actions can be done with it');
			}
			if (this._tabDriver.closed) {
				throw new Error('this tab has finished its work (close() was called), no other actions can be done with it');
			}
			if (this._actionInProgress) {
				throw new Error('cannot do this while another tab method is already running, each tab can execute only one action at a time');
			}
			var getAugmentedCallback = function getAugmentedCallback(callback) {
				var that = _this;
				return function () {
					that._actionInProgress = false;
					callback.apply(null, arguments);
				};
			};
			this._actionInProgress = true;
			if (callback != null) {
				if (typeof callback !== 'function') {
					throw new TypeError('callback parameter must be of type function');
				}
				action(getAugmentedCallback(callback));
			} else {
				return Promise.fromCallback(function (callback) {action(getAugmentedCallback(callback));}, { multiArgs: multiArgs });
			}
		} }, { key: "wait", value: function wait(

		duration) {var callback = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
			if (typeof duration !== 'number' || duration <= 0) {
				throw new TypeError("wait: duration parameter must be a positive number");
			}
			return this._callToTabDriver(function (callback) {
				setTimeout(function () {
					callback();
				}, duration);
			}, callback);
		} }, { key: "close", value: function close()

		{var _this2 = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			return this._callToTabDriver(function (callback) {
				_this2._tabDriver._close(function (err) {
					if (!err) {
						_this2.nick.unrefTabById(_this2.id);
					}
					callback(err);
				});
			}, callback);
		} }, { key: "open", value: function open(

		url) {var _this3 = this;var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
			if (typeof url !== 'string') {
				throw new TypeError('open: url parameter must be of type string');
			}
			if (typeof options === 'function') {
				callback = options;
				options = {};
			}
			if (!_.isPlainObject(options)) {
				throw new TypeError('open: options parameter must be of type plain object');
			}
			if (url.indexOf('://') < 0) {
				url = "http://" + url;
			}
			return this._callToTabDriver(function (callback) {_this3._tabDriver._open(url.trim(), options, callback);}, callback, true);
		} }, { key: "isVisible", value: function isVisible(

		selectors, operator, callback) {return this._isVisibleOrPresent("_waitUntilVisible", selectors, operator, callback);} }, { key: "isPresent", value: function isPresent(
		selectors, operator, callback) {return this._isVisibleOrPresent("_waitUntilPresent", selectors, operator, callback);} }, { key: "_isVisibleOrPresent", value: function _isVisibleOrPresent(
		method, selectors) {var _this4 = this;var operator = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;var callback = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
			if (typeof operator === 'function') {
				callback = operator;
				operator = null;
			}
			var f = function f(callback) {
				_this4._callTabDriverWaitMethod(method, selectors, 1, operator, function (err) {
					callback(null, !Boolean(err));
				});
			};
			if (callback) {
				f(callback);
			} else {
				return Promise.fromCallback(function (callback) {f(callback);});
			}
		} }, { key: "untilVisible", value: function untilVisible(


		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitUntilVisible', selectors, duration, operator, callback);} }, { key: "whileVisible", value: function whileVisible(
		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitWhileVisible', selectors, duration, operator, callback);} }, { key: "untilPresent", value: function untilPresent(
		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitUntilPresent', selectors, duration, operator, callback);} }, { key: "whilePresent", value: function whilePresent(
		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitWhilePresent', selectors, duration, operator, callback);} }, { key: "waitUntilVisible", value: function waitUntilVisible(

		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitUntilVisible', selectors, duration, operator, callback);} }, { key: "waitWhileVisible", value: function waitWhileVisible(
		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitWhileVisible', selectors, duration, operator, callback);} }, { key: "waitUntilPresent", value: function waitUntilPresent(
		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitUntilPresent', selectors, duration, operator, callback);} }, { key: "waitWhilePresent", value: function waitWhilePresent(
		selectors, duration, operator, callback) {return this._callTabDriverWaitMethod('_waitWhilePresent', selectors, duration, operator, callback);} }, { key: "_callTabDriverWaitMethod", value: function _callTabDriverWaitMethod(
		method, selectors) {var duration = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;var _this5 = this;var operator = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;var callback = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : null;
			if (typeof selectors === 'string') {
				selectors = [selectors];
			} else if (Array.isArray(selectors)) {
				if (selectors.length > 0) {var _iteratorNormalCompletion = true;var _didIteratorError = false;var _iteratorError = undefined;try {
						for (var _iterator = (0, _getIterator3.default)(selectors), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {var sel = _step.value;
							if (typeof sel !== 'string') {
								throw new TypeError(method + ": selectors parameter must be a string or an array of strings (css paths)");
							}
						}} catch (err) {_didIteratorError = true;_iteratorError = err;} finally {try {if (!_iteratorNormalCompletion && _iterator.return) {_iterator.return();}} finally {if (_didIteratorError) {throw _iteratorError;}}}} else

				throw new TypeError(method + ": selectors parameter must contain at least one string (css path)");
			} else {
				throw new TypeError(method + ": selectors parameter must be a string or an array of strings (css paths)");
			}
			selectors = selectors.map(function (s) {
				s = s.trim();
				if (s.length <= 0) {
					throw new TypeError(method + ": selectors parameter cannot contain an empty string");
				}
				return s;
			});
			if (typeof duration === 'function') {
				callback = duration;
				duration = null;
			} else if (operator === 'function') {
				callback = operator;
				operator = null;
			}
			if (typeof duration === 'string') {

				var d = duration;
				duration = operator;
				operator = d;
			}
			if (duration === null) {
				duration = 5000;
			} else if (typeof duration !== 'number' || duration <= 0) {
				throw new TypeError(method + ": duration parameter must be a positive number");
			}
			if (operator === null) {
				operator = 'and';
			} else if (operator !== 'and' && operator !== 'or') {
				throw new TypeError(method + ": operator parameter must be either \"and\" or \"or\"");
			}
			return this._callToTabDriver(function (callback) {_this5._tabDriver[method](selectors, duration, operator, callback);}, callback);
		} }, { key: "click", value: function click(

		selector) {var _this6 = this;var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
			if (typeof selector !== 'string') {
				throw new TypeError('click: selector parameter must be of type string');
			}
			if (typeof options === 'function') {
				callback = options;
				options = {};
			}
			if (!_.isPlainObject(options)) {

				throw new TypeError('click: options parameter must be of type plain object');
			}
			if (_.has(options, 'mouseEmulation')) {
				if (typeof options.submit !== 'boolean') {
					throw new TypeError('click: mouseEmulation option must be of type boolean');
				}
			} else {
				options.mouseEmulation = false;
			}


			return this._callToTabDriver(function (callback) {_this6._tabDriver._click(selector, options, callback);}, callback);

		} }, { key: "evaluate", value: function evaluate(

		func) {var _this7 = this;var arg = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
			if (_.isPlainObject(func)) {

				var f = func;
				func = arg;
				arg = f;
			}
			if (typeof func !== 'function') {
				throw new TypeError('evaluate: func parameter must be of type function');
			}
			if (typeof arg === 'function') {
				callback = arg;
				arg = null;
			}
			if (arg != null && !_.isPlainObject(arg) && !_.isArray(arg)) {
				throw new TypeError('evaluate: arg parameter must be a plain object or an array');
			}
			return this._callToTabDriver(function (callback) {_this7._tabDriver._evaluate(func, arg, callback);}, callback);
		} }, { key: "getUrl", value: function getUrl()

		{var _this8 = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			return this._callToTabDriver(function (callback) {_this8._tabDriver._getUrl(callback);}, callback);
		} }, { key: "getContent", value: function getContent()

		{var _this9 = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			return this._callToTabDriver(function (callback) {_this9._tabDriver._getContent(callback);}, callback);
		} }, { key: "fill", value: function fill(

		selector, params) {var _this10 = this;var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};var callback = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
			if (typeof selector !== 'string') {
				throw new TypeError('fill: selector parameter must be of type string');
			}
			if (!_.isPlainObject(params)) {
				throw new TypeError('fill: params parameter must be of type plain object');
			}

			if (typeof options === 'function') {
				callback = options;
				options = {};
			}
			if (!_.isPlainObject(options)) {
				throw new TypeError('fill: options parameter must be of type plain object');
			}
			if (_.has(options, 'submit')) {
				if (typeof options.submit !== 'boolean') {
					throw new TypeError('submit option must be of type boolean');
				}
			} else {
				options.submit = false;
			}


			return this._callToTabDriver(function (callback) {_this10._tabDriver._fill(selector, params, options, callback);}, callback);

		} }, { key: "screenshot", value: function screenshot(

		filename) {var _this11 = this;var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
			if (filename != null && typeof filename !== 'string') {
				throw new TypeError('screenshot: filename parameter must be null or of type string');
			}


			if (typeof options === 'function') {
				callback = options;
				options = {};
			}
			if (!_.isPlainObject(options)) {
				throw new TypeError('screenshot: options parameter must be of type plain object');
			}

			return this._callToTabDriver(function (callback) {_this11._tabDriver._screenshot(filename, options, callback);}, callback);
		} }, { key: "sendKeys", value: function sendKeys(

		selector, keys) {var _this12 = this;var options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};var callback = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
			if (typeof selector !== 'string') {
				throw new TypeError('sendKeys: selector parameter must be of type string');
			}
			if (typeof keys !== 'string' && typeof keys !== 'number') {
				throw new TypeError('sendKeys: keys parameter must be of type string or number');
			}
			if (typeof options === 'function') {
				callback = options;
				options = {};
			}
			if (!_.isPlainObject(options)) {
				throw new TypeError('sendKeys: options parameter must be of type plain object');
			}
			if (_.has(options, 'keepFocus')) {
				if (typeof options.keepFocus !== 'boolean') {
					throw new TypeError('sendKeys: keepFocus option must be of type boolean');
				}
			}
			if (_.has(options, 'reset')) {
				if (typeof options.reset !== 'boolean') {
					throw new TypeError('sendKeys: reset option must be of type boolean');
				}
			}
			if (_.has(options, 'modifiers')) {
				if (typeof options.reset !== 'string') {
					throw new TypeError('sendKeys: modifiers option must be of type string');
				}
			}


			return this._callToTabDriver(function (callback) {_this12._tabDriver._sendKeys(selector, keys, options, callback);}, callback);

		} }, { key: "inject", value: function inject(

		url) {var _this13 = this;var callback = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
			if (typeof url !== 'string') {
				throw new TypeError('inject: url parameter must be of type string');
			}

			if (url.trim().toLowerCase().indexOf('http://') === 0 || url.trim().toLowerCase().indexOf('https://') === 0 || url.trim().toLowerCase().indexOf('file://') === 0) {
				return this._callToTabDriver(function (callback) {_this13._tabDriver._injectFromUrl(url, callback);}, callback);
			} else {
				return this._callToTabDriver(function (callback) {_this13._tabDriver._injectFromDisk(url, callback);}, callback);
			}
		} }, { key: "scrollTo", value: function scrollTo(

		x, y, callback) {
			return this.scroll(x, y, callback);
		} }, { key: "scroll", value: function scroll(

		x, y) {var _this14 = this;var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
			if (typeof x !== "number") {
				throw new TypeError("scrollTo: x parameter must be of type number");
			}
			if (typeof y !== "number") {
				throw new TypeError("scrollTo: y parameter must be of type number");
			}
			return this._callToTabDriver(function (callback) {_this14._tabDriver._scroll(x, y, callback);}, callback);
		} }, { key: "scrollToBottom", value: function scrollToBottom()

		{var _this15 = this;var callback = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : null;
			return this._callToTabDriver(function (callback) {_this15._tabDriver._scrollToBottom(callback);}, callback);
		} }, { key: "getAllCookies", value: function getAllCookies(




		callback) {return this.nick.getAllCookies(callback);} }, { key: "deleteAllCookies", value: function deleteAllCookies(
		callback) {return this.nick.deleteAllCookies(callback);} }, { key: "deleteCookie", value: function deleteCookie(
		name, domain, callback) {return this.nick.deleteCookie(name, domain, callback);} }, { key: "setCookie", value: function setCookie(
		cookie, callback) {return this.nick.setCookie(cookie, callback);} }, { key: "nick", get: function get() {return this._nick;} }, { key: "driver", get: function get() {return this._tabDriver;} }, { key: "tabDriver", get: function get() {return this._tabDriver;} }, { key: "actionInProgress", get: function get() {return this._actionInProgress;} }, { key: "closed", get: function get() {return this._tabDriver.closed;} }, { key: "crashed", get: function get() {return this._tabDriver.crashed;} }, { key: "id", get: function get() {return this._tabDriver.id;} }, { key: "onConfirm", set: function set(f) {if (typeof f !== "function") {throw new TypeError("onConfirm must receive a function");}this._tabDriver._onConfirm = f;} }, { key: "onPrompt", set: function set(f) {if (typeof f !== "function") {throw new TypeError("onPrompt must receive a function");}this._tabDriver._onPrompt = f;} }]);return Tab;}();



module.exports = Tab;
