// nickjs/lib/es5/chrome/TabDriver.js
'use strict';var _from = require('babel-runtime/core-js/array/from');var _from2 = _interopRequireDefault(_from);var _typeof2 = require('babel-runtime/helpers/typeof');var _typeof3 = _interopRequireDefault(_typeof2);var _stringify = require('babel-runtime/core-js/json/stringify');var _stringify2 = _interopRequireDefault(_stringify);var _getIterator2 = require('babel-runtime/core-js/get-iterator');var _getIterator3 = _interopRequireDefault(_getIterator2);var _promise = require('babel-runtime/core-js/promise');var _promise2 = _interopRequireDefault(_promise);var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require('babel-runtime/helpers/createClass');var _createClass3 = _interopRequireDefault(_createClass2);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}



var _ = require('lodash');var

TabDriver = function () {

	function TabDriver(uniqueTabId, options, client, cdpTargetId) {(0, _classCallCheck3.default)(this, TabDriver);
		this.__closed = false;
		this.__crashed = false;
		this.__uniqueTabId = uniqueTabId;
		this.__options = options;
		this.__client = client;
		this.__cdpTargetId = cdpTargetId;


		this._onConfirm = null;
		this._onPrompt = null;
	}(0, _createClass3.default)(TabDriver, [{ key: '_init', value: function _init(



		callback) {var _this = this;
			_promise2.default.all([
			this.__client.Page.enable(),
			this.__client.Network.enable(),
			this.__client.Runtime.enable(),
			this.__client.Security.enable()]).
			then(function () {


				return _this.__client.Security.setOverrideCertificateErrors({ override: true }).then(function () {
					_this.__client.Security.certificateError(function (e) {
						var payload = {
							eventId: e.eventId,
							action: 'continue' };

						_this.__client.Security.handleCertificateError(payload, function (err, res) {
							err = _this.__parseCdpResponse(err, res, "handleCertificateError failure");
							if (err) {
								console.log('> Tab ' + _this.id + ': ' + err);
							}
						});
					});
				});

			}).then(function () {


				if (_this.__options.whitelist.length > 0 ||
				_this.__options.blacklist.length > 0 ||
				_this.__options._proxyUsername ||
				_this.__options._proxyPassword ||
				_this.__options.loadImages === false) {
					return _this.__client.Network.setRequestInterception({ patterns: [{ urlPattern: "*" }] }).then(function () {
						var processRequest = function processRequest(e, credentials, error, logMessage) {
							if (logMessage && _this.__options.printAborts) {
								console.log('> Tab ' + _this.id + ': Aborted (' + logMessage + '): ' + e.request.url);
							}
							var payload = {
								interceptionId: e.interceptionId };

							if (error) {
								payload.errorReason = error;
							}
							if (credentials) {
								payload.authChallengeResponse = credentials;
							}
							_this.__client.Network.continueInterceptedRequest(payload, function (err, res) {
								err = _this.__parseCdpResponse(err, res, "continueInterceptedRequest failure");

								if (err && err.indexOf("Invalid InterceptionId.") < 0) {
									console.log('> Tab ' + _this.id + ': ' + err);
								}
							});
						};
						_this.__client.Network.requestIntercepted(function (e) {
							var credentials = null;
							if (e.authChallenge && e.authChallenge.source === "Proxy" && (_this.__options._proxyUsername || _this.__options._proxyPassword)) {
								credentials = {
									response: "ProvideCredentials",
									username: _this.__options._proxyUsername,
									password: _this.__options._proxyPassword };

							}
							if (_this.__options.loadImages === false && e.resourceType === "Image") {
								return processRequest(e, credentials, "Aborted");
							}
							if (_this.__options.whitelist.length > 0) {
								var found = false;var _iteratorNormalCompletion = true;var _didIteratorError = false;var _iteratorError = undefined;try {
									for (var _iterator = (0, _getIterator3.default)(_this.__options.whitelist), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {var white = _step.value;
										if (typeof white === "string") {
											var url = e.request.url.toLowerCase();
											if (url.indexOf(white) === 0 || url.indexOf('https://' + white) === 0 || url.indexOf('http://' + white) === 0) {
												found = true;
												break;
											}
										} else if (white.test(e.request.url)) {
											found = true;
											break;
										}
									}} catch (err) {_didIteratorError = true;_iteratorError = err;} finally {try {if (!_iteratorNormalCompletion && _iterator.return) {_iterator.return();}} finally {if (_didIteratorError) {throw _iteratorError;}}}
								if (!found) {
									return processRequest(e, credentials, "Aborted", "not found in whitelist");
								}
							}var _iteratorNormalCompletion2 = true;var _didIteratorError2 = false;var _iteratorError2 = undefined;try {
								for (var _iterator2 = (0, _getIterator3.default)(_this.__options.blacklist), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {var black = _step2.value;
									if (typeof black === 'string') {
										var _url = e.request.url.toLowerCase();
										if (_url.indexOf(black) === 0 || _url.indexOf('https://' + black) === 0 || _url.indexOf('http://' + black) === 0) {
											return processRequest(e, credentials, "Aborted", 'blacklisted by "' + black + '"');
										}
									} else if (black.test(e.request.url)) {
										return processRequest(e, credentials, "Aborted", 'blacklisted by ' + black);
									}
								}} catch (err) {_didIteratorError2 = true;_iteratorError2 = err;} finally {try {if (!_iteratorNormalCompletion2 && _iterator2.return) {_iterator2.return();}} finally {if (_didIteratorError2) {throw _iteratorError2;}}}
							processRequest(e, credentials);
						});
					});
				}

			}).then(function () {
				return _this.__client.Network.setUserAgentOverride({ userAgent: _this.__options.userAgent });
			}).then(function () {





















				if (_this.__options.printPageErrors) {

					_this.__client.Runtime.exceptionThrown(function (e) {
						console.log('> Tab ' + _this.id + ': Page JavaScript error: ' + _this.__summarizeException(e));
					});
				}

				if (_this.__options.printNavigation) {
					_this.__client.Page.frameScheduledNavigation(function (e) {
						if (e.reason && e.url) {
							console.log('> Tab ' + _this.id + ': Navigation (' + e.reason + '): ' + e.url);
						}
					});
				}















				_this.__client.Page.javascriptDialogOpening(function (e) {
					var payload = {};
					if (e.type === "confirm") {
						payload.accept = _this._onConfirm(e.message);
					} else if (e.type === "prompt") {
						var text = _this._onPrompt(e.message);
						if (typeof text === "string") {
							payload.accept = true;
							payload.promptText = text;
						} else {
							payload.accept = false;
							payload.promptText = "";
						}
					} else {
						payload.accept = false;
					}
					_this.__client.Page.handleJavaScriptDialog(payload, function (err, res) {
						err = _this.__parseCdpResponse(err, res, "handleJavaScriptDialog failure");
						if (err) {
							console.log('> Tab ' + _this.id + ': ' + err);
						}
					});
				});

				var setupNickJsInPage = function setupNickJsInPage() {
					window.__nativePromise = _promise2.default;
				};
				_this.__client.Page.addScriptToEvaluateOnNewDocument({
					source: '(' + setupNickJsInPage + ')()' },
				function (err, res) {
					err = _this.__parseCdpResponse(err, res, "addScriptToEvaluateOnNewDocument failure");
					if (err) {
						console.log('> Tab ' + _this.id + ': ' + err);
					}
				});

			}).then(function () {
				callback(null);
			}).catch(function (err) {
				callback('error when initializing new chrome tab: ' + err);
			});
		} }, { key: '__summarizeException', value: function __summarizeException(

		e) {
			if (this.__options.debug) {
				console.log((0, _stringify2.default)(e, undefined, 8));
			}
			if (_.isPlainObject(e.exceptionDetails.exception) && typeof e.exceptionDetails.exception.description === "string") {

				return e.exceptionDetails.exception.description;
			} else if (_.isPlainObject(e.exceptionDetails.exception) && typeof e.exceptionDetails.exception.value === "string") {

				return e.exceptionDetails.exception.value;
			} else {
				if (typeof e.exceptionDetails.text === "string") {

					return e.exceptionDetails.text;
				} else {

					return "uncaught exception";
				}
			}
		} }, { key: '__parseCdpResponse', value: function __parseCdpResponse(

		err, res, errorText) {
			if (err === true && _.isPlainObject(res)) {

				return errorText + ': DevTools protocol error: ' + (typeof res.message === "string" ? res.message : "unknown error") + (typeof res.code === "number" ? ' (code: ' + res.code + ')' : "");
			} else if (_.isPlainObject(res) && _.isPlainObject(res.exceptionDetails)) {

				return errorText + ': exception thrown from page: ' + this.__summarizeException(res);
			} else if (err) {

				return errorText + ': ' + err;
			} else {

				return null;
			}
		} }, { key: '_chromeHasCrashed', value: function _chromeHasCrashed()








		{
			this.__crashed = true;
		} }, { key: '_close', value: function _close(

		callback) {var _this2 = this;
			this.__client.close(function () {
				var CDP = require("chrome-remote-interface");
				CDP.Close({ id: _this2.__cdpTargetId }, function (err, res) {
					_this2.__closed = true;
					callback(_this2.__parseCdpResponse(err, res, "failed to close chrome tab"));
				});
			});
		} }, { key: '_open', value: function _open(

		url, options, callback) {var _this3 = this;

			var requestId = null;


			var status = null;
			var statusText = null;
			var newUrl = null;


			var callbackHasFired = false;
			var fireCallback = function fireCallback(err) {
				if (!callbackHasFired) {
					callbackHasFired = true;
					_this3.__client.removeListener("Network.responseReceived", responseReceived);
					_this3.__client.removeListener("Network.requestWillBeSent", requestWillBeSent);
					_this3.__client.removeListener("Page.loadEventFired", loadEventFired);
					_this3.__client.removeListener("Network.loadingFailed", loadingFailed);
					clearTimeout(timeoutId);
					callback(err, status, statusText, newUrl);
				}
			};


			var responseReceived = function responseReceived(e) {
				if (e.requestId === requestId && status === null) {
					status = e.response.status;
					statusText = e.response.statusText;
					newUrl = e.response.url;
				}
			};


			var requestWillBeSent = function requestWillBeSent(e) {


				if (status === null) {
					if (e.documentURL === url || e.documentURL === url + '/') {
						requestId = e.requestId;
					}
				}
			};



			var loadEventFired = function loadEventFired() {
				if (_this3.__options.printNavigation) {
					console.log('> Tab ' + _this3.id + ': Navigation (open): ' + (newUrl || url));
				}
				fireCallback(null);
			};


			var loadingFailed = function loadingFailed(e) {
				if (e.requestId === requestId) {
					var errorMessage = e.errorText || "unknown error";
					if (_this3.__options.printNavigation) {
						console.log('> Tab ' + _this3.id + ': Navigation (open error): ' + errorMessage + ' (' + (newUrl || url) + ')');
					}
					fireCallback('loading failed: ' + errorMessage);
				}
			};


			this.__client.on("Network.responseReceived", responseReceived);
			this.__client.on("Network.requestWillBeSent", requestWillBeSent);
			this.__client.on("Page.loadEventFired", loadEventFired);
			this.__client.on("Network.loadingFailed", loadingFailed);


			var onTimeout = function onTimeout() {
				var errorMessage = 'load event did not fire after ' + (Date.now() - openStart) + 'ms';
				if (_this3.__options.printNavigation) {
					console.log('> Tab ' + _this3.id + ': Navigation (open timeout): ' + errorMessage + ' (' + (newUrl || url) + ')');
				}
				fireCallback('timeout: ' + errorMessage);
			};
			var timeoutId = setTimeout(onTimeout, this.__options.timeout);
			var openStart = Date.now();


			this.__client.Page.navigate({ url: url }, function (err, res) {
				err = _this3.__parseCdpResponse(err, res, "failed to make chrome navigate");
				if (err) {
					fireCallback(err);
				}
			});
		} }, { key: '_waitUntilVisible', value: function _waitUntilVisible(

		selectors, duration, operator, callback) {this.__callWaitMethod('until', 'visible', selectors, duration, operator, callback);} }, { key: '_waitWhileVisible', value: function _waitWhileVisible(
		selectors, duration, operator, callback) {this.__callWaitMethod('while', 'visible', selectors, duration, operator, callback);} }, { key: '_waitUntilPresent', value: function _waitUntilPresent(
		selectors, duration, operator, callback) {this.__callWaitMethod('until', 'present', selectors, duration, operator, callback);} }, { key: '_waitWhilePresent', value: function _waitWhilePresent(
		selectors, duration, operator, callback) {this.__callWaitMethod('while', 'present', selectors, duration, operator, callback);} }, { key: '__callWaitMethod', value: function __callWaitMethod(
		waitType, visType, selectors, duration, operator, callback) {var _this4 = this;
			var waiterToInject = function waiterToInject(waitType, visType, selectors, timeLeft, timeSpent, operator) {
				if (visType === 'visible') {
					var selectorMatches = function selectorMatches(selector) {
						var ret = false;var _iteratorNormalCompletion3 = true;var _didIteratorError3 = false;var _iteratorError3 = undefined;try {
							for (var _iterator3 = (0, _getIterator3.default)(document.querySelectorAll(selector)), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {var element = _step3.value;
								var style = window.getComputedStyle(element);
								if (style.visibility !== 'hidden' && style.display !== 'none') {
									var rect = element.getBoundingClientRect();
									if (rect.width > 0 && rect.height > 0) {
										ret = true;
										break;
									}
								}
							}} catch (err) {_didIteratorError3 = true;_iteratorError3 = err;} finally {try {if (!_iteratorNormalCompletion3 && _iterator3.return) {_iterator3.return();}} finally {if (_didIteratorError3) {throw _iteratorError3;}}}
						return waitType === 'while' ? !ret : ret;
					};
				} else {
					var selectorMatches = function selectorMatches(selector) {
						var ret = document.querySelector(selector) != null;
						return waitType === 'while' ? !ret : ret;
					};
				}
				return new window.__nativePromise(function (fulfill, reject) {
					var start = Date.now();
					if (operator === "and") {
						var waitForAll = function waitForAll() {
							var invalidSelector = null;var _iteratorNormalCompletion4 = true;var _didIteratorError4 = false;var _iteratorError4 = undefined;try {
								for (var _iterator4 = (0, _getIterator3.default)(selectors), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {var sel = _step4.value;
									if (!selectorMatches(sel)) {

										invalidSelector = sel;
										break;
									}
								}} catch (err) {_didIteratorError4 = true;_iteratorError4 = err;} finally {try {if (!_iteratorNormalCompletion4 && _iterator4.return) {_iterator4.return();}} finally {if (_didIteratorError4) {throw _iteratorError4;}}}
							if (invalidSelector) {
								if (start + timeLeft < Date.now()) {
									reject('waited ' + (timeSpent + (Date.now() - start)) + 'ms but element "' + invalidSelector + '" still ' + (waitType === 'while' ? '' : 'not ') + visType);
								} else {
									setTimeout(function () {return waitForAll();}, 50);
								}
							} else {

								fulfill();
							}
						};
						waitForAll();
					} else {
						var waitForOne = function waitForOne() {
							var firstMatch = null;var _iteratorNormalCompletion5 = true;var _didIteratorError5 = false;var _iteratorError5 = undefined;try {
								for (var _iterator5 = (0, _getIterator3.default)(selectors), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {var sel = _step5.value;
									if (selectorMatches(sel)) {
										firstMatch = sel;

										break;
									}
								}} catch (err) {_didIteratorError5 = true;_iteratorError5 = err;} finally {try {if (!_iteratorNormalCompletion5 && _iterator5.return) {_iterator5.return();}} finally {if (_didIteratorError5) {throw _iteratorError5;}}}
							if (firstMatch) {
								fulfill(firstMatch);
							} else {
								if (start + timeLeft < Date.now()) {
									var elementsToString = selectors.slice();var _iteratorNormalCompletion6 = true;var _didIteratorError6 = false;var _iteratorError6 = undefined;try {
										for (var _iterator6 = (0, _getIterator3.default)(elementsToString), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {var e = _step6.value;
											e = '"' + e + '"';
										}} catch (err) {_didIteratorError6 = true;_iteratorError6 = err;} finally {try {if (!_iteratorNormalCompletion6 && _iterator6.return) {_iterator6.return();}} finally {if (_didIteratorError6) {throw _iteratorError6;}}}
									elementsToString = elementsToString.join(', ');
									reject('waited ' + (timeSpent + (Date.now() - start)) + 'ms but element' + (selectors.length > 0 ? 's' : '') + ' ' + elementsToString + ' still ' + (waitType === 'while' ? '' : 'not ') + visType);
								} else {

									setTimeout(function () {return waitForOne();}, 50);
								}
							}
						};
						waitForOne();
					}
				});
			};
			var tryToWait = function tryToWait(timeLeft, timeSpent) {

				var payload = {
					expression: '(' + waiterToInject + ')("' + waitType + '", "' + visType + '", ' + (0, _stringify2.default)(selectors) + ', ' + timeLeft + ', ' + timeSpent + ', "' + operator + '")',
					awaitPromise: true,
					includeCommandLineAPI: false };

				var start = Date.now();
				_this4.__client.Runtime.evaluate(payload, function (err, res) {
					if (_.has(res, "message") && (res.message === "Promise was collected" || res.message === "Execution context was destroyed.")) {



						var timeElapsed = Date.now() - start;
						tryToWait(timeLeft - timeElapsed, timeSpent + timeElapsed);
					} else {
						err = _this4.__parseCdpResponse(err, res, 'wait ' + waitType + ' ' + visType + ' failure');
						if (err) {
							callback(err);
						} else {
							if (res && (0, _typeof3.default)(res.result) === "object" && typeof res.result.value === "string") {
								callback(null, res.result.value);
							} else {
								callback(null, null);
							}
						}
					}
				});
			};
			tryToWait(duration, 0);
		} }, { key: '_click', value: function _click(

		selector, options, callback) {var _this5 = this;

			var click = function click(selector) {
				var target = document.querySelector(selector);
				if (target) {

					var posX = 0.5;
					var posY = 0.5;
					try {
						var bounds = target.getBoundingClientRect();
						posX = Math.floor(bounds.width * (posX - (posX ^ 0)).toFixed(10)) + (posX ^ 0) + bounds.left;
						posY = Math.floor(bounds.height * (posY - (posY ^ 0)).toFixed(10)) + (posY ^ 0) + bounds.top;
					} catch (e) {
						posX = 1;
						posY = 1;
					}
					target.dispatchEvent(new MouseEvent("click", {
						bubbles: true,
						cancelable: true,
						view: window,
						detail: 1,
						screenX: 1,
						screenY: 1,
						clientX: posX,
						clientY: posY,
						ctrlKey: false,
						altKey: false,
						shiftKey: false,
						metaKey: false,
						button: 0,
						relatedTarget: target }));

				} else {
					throw 'cannot find element "' + selector + '"';
				}
			};
			var payload = {
				expression: '(' + click + ')(' + (0, _stringify2.default)(selector) + ')',
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				callback(_this5.__parseCdpResponse(err, res, 'click: failed to click on target element "' + selector + '"'));
			});
		} }, { key: '_evaluate', value: function _evaluate(

		func, arg, callback) {var _this6 = this;
			var runEval = function runEval(func, arg) {
				return new window.__nativePromise(function (fulfill, reject) {
					__done = function __done(err, res) {
						if (err) {
							reject(err);
						} else {
							fulfill(res);
						}
					};

					func(arg, __done);
				});
			};
			var payload = {
				expression: '(' + runEval + ')((' + func + '), ' + (0, _stringify2.default)(arg) + ')',
				awaitPromise: true,
				returnByValue: true,
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				err = _this6.__parseCdpResponse(err, res, "evaluate: code evaluation failed");
				if (err) {
					callback(err, null);
				} else {

					callback(null, res.result.value);
				}
			});
		} }, { key: '_getUrl', value: function _getUrl(

		callback) {var _this7 = this;
			var payload = {
				expression: "window.location.href",
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				err = _this7.__parseCdpResponse(err, res, "getUrl: could not get the current url");
				if (err) {
					callback(err);
				} else {
					callback(null, res.result.value);
				}
			});
		} }, { key: '_getContent', value: function _getContent(

		callback) {var _this8 = this;
			this.__client.DOM.getDocument(function (err, res) {
				err = _this8.__parseCdpResponse(err, res, "getContent: failed to get root dom node from page");
				if (err) {
					callback(err);
				} else {
					_this8.__client.DOM.getOuterHTML({ nodeId: res.root.nodeId }, function (err, res) {
						err = _this8.__parseCdpResponse(err, res, "getContent: failed to get outer html from root dom node");
						if (err) {
							callback(err);
						} else {
							callback(null, res.outerHTML);
						}
					});
				}
			});
		} }, { key: '_fill', value: function _fill(

		selector, params, options, callback) {var _this9 = this;


			var fillForm = function fillForm(selector, params, options) {
				var form = document.querySelector(selector);
				if (!form) {
					throw 'cannot find any element matching "' + selector + '"';
				} else if (form.nodeName !== "FORM") {
					if (typeof form.nodeName === "string") {
						throw 'element "' + selector + '" is a ' + form.nodeName + ', not a form';
					} else {
						throw 'element "' + selector + '" is not a form';
					}
				}
				for (var inputName in params) {
					var desiredValues = params[inputName];
					if (!Array.isArray(desiredValues)) {
						desiredValues = [desiredValues];
					}
					var matchingFields = form.querySelectorAll('[name="' + inputName + '"]');var _iteratorNormalCompletion7 = true;var _didIteratorError7 = false;var _iteratorError7 = undefined;try {
						for (var _iterator7 = (0, _getIterator3.default)(matchingFields), _step7; !(_iteratorNormalCompletion7 = (_step7 = _iterator7.next()).done); _iteratorNormalCompletion7 = true) {var field = _step7.value;
							try {
								field.focus();
							} catch (e) {}
							if (field.getAttribute("contenteditable")) {
								field.textContent = desiredValues[0];
							} else {
								var fieldType = field.nodeName.toLowerCase();
								if (field.getAttribute("type")) {
									fieldType = field.getAttribute("type").toLowerCase();
								}
								switch (fieldType) {
									case "checkbox":
									case "radio":


										if (desiredValues.length === 1 && typeof desiredValues[0] === "boolean") {
											field.checked = desiredValues[0];
										} else {
											field.checked = desiredValues.indexOf(field.getAttribute("value")) >= 0;
										}
										break;
									case "file":
										throw "inputs of type \"file\" are not handled yet, sorry";
									case "select":var _iteratorNormalCompletion8 = true;var _didIteratorError8 = false;var _iteratorError8 = undefined;try {
											for (var _iterator8 = (0, _getIterator3.default)(field.options), _step8; !(_iteratorNormalCompletion8 = (_step8 = _iterator8.next()).done); _iteratorNormalCompletion8 = true) {var option = _step8.value;
												option.selected = desiredValues.indexOf(option.value) >= 0;
											}} catch (err) {_didIteratorError8 = true;_iteratorError8 = err;} finally {try {if (!_iteratorNormalCompletion8 && _iterator8.return) {_iterator8.return();}} finally {if (_didIteratorError8) {throw _iteratorError8;}}}

										if (!field.value || !field.multiple && field.value !== desiredValues[0]) {var _iteratorNormalCompletion9 = true;var _didIteratorError9 = false;var _iteratorError9 = undefined;try {
												for (var _iterator9 = (0, _getIterator3.default)(field.options), _step9; !(_iteratorNormalCompletion9 = (_step9 = _iterator9.next()).done); _iteratorNormalCompletion9 = true) {var _option = _step9.value;
													_option.selected = desiredValues.indexOf(_option.text) >= 0;
												}} catch (err) {_didIteratorError9 = true;_iteratorError9 = err;} finally {try {if (!_iteratorNormalCompletion9 && _iterator9.return) {_iterator9.return();}} finally {if (_didIteratorError9) {throw _iteratorError9;}}}
										}
										break;
									default:
										field.value = desiredValues[0];}

							}
							field.dispatchEvent(new Event("change", { bubbles: true, cancelable: true }));
							field.dispatchEvent(new Event("input", { bubbles: true, cancelable: true }));
							try {
								field.blur();
							} catch (e) {}
						}} catch (err) {_didIteratorError7 = true;_iteratorError7 = err;} finally {try {if (!_iteratorNormalCompletion7 && _iterator7.return) {_iterator7.return();}} finally {if (_didIteratorError7) {throw _iteratorError7;}}}
				}
				if (options.submit) {
					if (!form.dispatchEvent(new Event("submit", { bubbles: true, cancelable: true }))) {
						throw 'could not dispatch submit event to form "' + selector + '"';
					}
					if (typeof form.submit === "function") {
						form.submit();
					} else {


						document.createElement("form").submit.call(form);
					}
				}
			};
			var payload = {
				expression: '(' + fillForm + ')(' + (0, _stringify2.default)(selector) + ', ' + (0, _stringify2.default)(params) + ', ' + (0, _stringify2.default)(options) + ')',
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				callback(_this9.__parseCdpResponse(err, res, 'fill: error when filling "' + selector + '" form'));
			});
		} }, { key: '_screenshot', value: function _screenshot(

		filename, options, callback) {var _this10 = this;



			var ext = require("path").extname(filename).toLowerCase();
			var format = ext === ".png" ? "png" : "jpeg";
			var payload = {
				format: format };


			if (format === "jpeg") {
				payload.quality = options.quality || 75;
			}
			var takeScreenshot = function takeScreenshot(override) {
				var writeToDisk = function writeToDisk(err, res) {

					err = _this10.__parseCdpResponse(err, res, "screenshot: could not capture area");
					if (err) {
						callback(err);
					} else {
						require("fs").writeFile(filename, res.data, "base64", function (err) {
							if (err) {
								callback('screenshot: could not write captured data to disk: ' + err, filename);
							} else {
								callback(null, filename);
							}
						});
					}
				};
				_this10.__client.Page.captureScreenshot(payload, function (err, res) {
					if (override) {
						override.width = _this10.__options.width;
						override.height = _this10.__options.height;
						_this10.__client.Emulation.setDeviceMetricsOverride(override, function () {


							writeToDisk(err, res);
						});
					} else {
						writeToDisk(err, res);
					}
				});
			};
			if (typeof options.fullPage === 'boolean' && !options.fullPage) {
				takeScreenshot();
			} else {
				this.__client.Page.getLayoutMetrics({}, function (err, res) {
					err = _this10.__parseCdpResponse(err, res, "screenshot: could not get layout metrics");
					if (err) {
						callback(err);
					} else {
						var override = {
							mobile: false,
							width: Math.ceil(res.contentSize.width),
							height: Math.ceil(res.contentSize.height),
							deviceScaleFactor: 1,
							screenOrientation: {
								angle: 0,
								type: "portraitPrimary" } };


						_this10.__client.Emulation.setDeviceMetricsOverride(override, function (err, res) {
							err = _this10.__parseCdpResponse(err, res, "screenshot: could not resize viewport to full page");
							if (err) {
								callback(err);
							} else {
								takeScreenshot(override);
							}
						});
					}
				});
			}
		} }, { key: '_sendKeys', value: function _sendKeys(

		selector, keys, options, callback) {var _this11 = this;
			var focusElement = function focusElement(selector) {
				var target = document.querySelector(selector);
				if (target) {
					target.focus();
				} else {
					throw 'cannot find element "' + selector + '"';
				}
			};
			var blurElement = function blurElement(selector) {
				var target = document.querySelector(selector);
				if (target) {
					try {
						target.blur();
					} catch (e) {}
				}
			};
			var payload = {
				expression: '(' + focusElement + ')(' + (0, _stringify2.default)(selector) + ')',
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				err = _this11.__parseCdpResponse(err, res, 'sendKeys: could not focus editable element "' + selector + '"');
				if (err) {
					callback(err);
				} else {
					var dispatch = function dispatch(chr, type, callback) {
						payload = {
							type: type,
							text: chr };

						_this11.__client.Input.dispatchKeyEvent(payload, function (err, res) {

							callback(_this11.__parseCdpResponse(err, res, "sendKeys: could not dispatch key event"));
						});
					};
					var chrIterator = function chrIterator(chr, callback) {
						dispatch(chr, "keyDown", function (err) {
							if (err) {
								callback(err);
							} else {
								dispatch(chr, "keyUp", callback);
							}
						});
					};
					require("async").eachSeries((0, _from2.default)(keys), chrIterator, function (err) {
						if (err) {
							callback(err);
						} else {
							if (options.keepFocus) {
								callback(null);
							} else {
								payload = {
									expression: '(' + blurElement + ')(' + (0, _stringify2.default)(selector) + ')',
									includeCommandLineAPI: false };

								_this11.__client.Runtime.evaluate(payload, function (err, res) {
									callback(_this11.__parseCdpResponse(err, res, 'sendKeys: could not blur editable element "' + selector + '"'));
								});
							}
						}
					});
				}
			});
		} }, { key: '_injectFromDisk', value: function _injectFromDisk(

		url, callback) {var _this12 = this;
			require("fs").readFile(url, "utf8", function (err, data) {
				if (err) {
					callback('inject: could not read file from disk for injection into page: ' + err);
				} else {
					_this12.__injectString(data, callback);
				}
			});
		} }, { key: '_injectFromUrl', value: function _injectFromUrl(

		url, callback) {var _this13 = this;
			var options = {
				json: false,
				parse_response: false,
				parse_cookies: false };


			require("needle").get(url, options, function (err, res, data) {
				if (err) {
					callback('could not download file from url for injection into page: ' + err);
				} else {
					if (res.statusCode >= 200 && res.statusCode < 300) {
						_this13.__injectString(data.toString(), callback);
					} else {
						callback('inject: could not download file from url for injection into page: got HTTP ' + res.statusCode + ' ' + res.statusMessage);
					}
				}
			});
		} }, { key: '__injectString', value: function __injectString(

		str, callback) {var _this14 = this;
			var payload = {
				expression: str,
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				callback(_this14.__parseCdpResponse(err, res, "inject: could not inject script into page"));
			});
		} }, { key: '_scroll', value: function _scroll(

		x, y, callback) {var _this15 = this;
			var doScroll = function doScroll(x, y) {
				window.scroll(x, y);
			};
			var payload = {
				expression: '(' + doScroll + ')(' + x + ', ' + y + ')',
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				callback(_this15.__parseCdpResponse(err, res, 'scroll: could not scroll'));
			});
		} }, { key: '_scrollToBottom', value: function _scrollToBottom(

		callback) {var _this16 = this;
			var doScroll = function doScroll() {
				window.scroll(0, document.body.scrollHeight);
			};
			var payload = {
				expression: '(' + doScroll + ')()',
				includeCommandLineAPI: false };

			this.__client.Runtime.evaluate(payload, function (err, res) {
				callback(_this16.__parseCdpResponse(err, res, 'scrollToBottom: could not scroll to bottom'));
			});
		} }, { key: '_getAllCookies', value: function _getAllCookies(



		callback) {var _this17 = this;
			this.__client.Network.getAllCookies({}, function (err, res) {
				err = _this17.__parseCdpResponse(err, res, "getAllCookies: failed to get all cookies");
				if (err) {
					callback(err);
				} else {
					callback(null, res.cookies);
				}
			});
		} }, { key: '_deleteAllCookies', value: function _deleteAllCookies(
		callback) {var _this18 = this;
			this.__client.Network.clearBrowserCookies({}, function (err, res) {
				callback(_this18.__parseCdpResponse(err, res, "deleteAllCookies: failed to clear all cookies"));
			});
		} }, { key: '_deleteCookie', value: function _deleteCookie(
		name, domain, callback) {var _this19 = this;


			if (domain.toLowerCase().indexOf("http://") !== 0 && domain.toLowerCase().indexOf("https://") !== 0) {
				domain = 'http://' + domain;
			}
			var payload = {
				name: name,
				url: domain };

			this.__client.send("Network.deleteCookies", payload, function (err, res) {
				callback(_this19.__parseCdpResponse(err, res, 'deleteCookie: failed to delete cookie "' + name + '"'));
			});
		} }, { key: '_setCookie', value: function _setCookie(
		cookie, callback) {var _this20 = this;


			var url = cookie.domain;
			if (url.toLowerCase().indexOf("http://") !== 0 && url.toLowerCase().indexOf("https://") !== 0) {
				url = 'http://' + url;
			}
			var payload = {
				name: cookie.name,
				value: cookie.value,
				url: url,
				expirationDate: Math.round(Date.now() / 1000) + 365 * 24 * 60 * 60 };

			if (_.has(cookie, "secure")) {
				payload.secure = cookie.secure;
			}
			if (_.has(cookie, "httpOnly")) {
				payload.httpOnly = cookie.httpOnly;
			}
			this.__client.Network.setCookie(payload, function (err, res) {
				err = _this20.__parseCdpResponse(err, res, 'setCookie: could not add cookie "' + cookie.name + '"');
				if (err) {
					callback(err);
				} else {
					if (res.success === true) {
						callback(null);
					} else {
						callback('setCookie: could not add cookie "' + cookie.name + '"');
					}
				}
			});
		} }, { key: 'client', get: function get() {return this.__client;} }, { key: 'closed', get: function get() {return this.__closed;} }, { key: 'crashed', get: function get() {return this.__crashed;} }, { key: 'id', get: function get() {return this.__uniqueTabId;} }]);return TabDriver;}();



module.exports = TabDriver;
