// nickjs/lib/es5/casper/TabDriver.js
var require = patchRequire(require);"use strict";var _getIterator2 = require("babel-runtime/core-js/get-iterator");var _getIterator3 = _interopRequireDefault(_getIterator2);var _classCallCheck2 = require("babel-runtime/helpers/classCallCheck");var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);var _createClass2 = require("babel-runtime/helpers/createClass");var _createClass3 = _interopRequireDefault(_createClass2);function _interopRequireDefault(obj) {return obj && obj.__esModule ? obj : { default: obj };}







var _ = require("lodash");
var casper = require("casper");var

TabDriver = function () {

	function TabDriver(uniqueTabId, options) {var _this = this;(0, _classCallCheck3.default)(this, TabDriver);
		this.__uniqueTabId = uniqueTabId;
		this.__closed = false;
		this.__crashed = false;
		this.__endCallback = null;
		this.__nextStep = null;


		this._onConfirm = null;
		this._onPrompt = null;

		var casperOptions = {
			verbose: false,
			colorizerType: 'Dummy',
			exitOnError: true,
			silentErrors: false,
			retryTimeout: 25,
			pageSettings: {
				localToRemoteUrlAccessEnabled: true,
				webSecurityEnabled: false,
				loadPlugins: false,
				userAgent: options.userAgent,
				resourceTimeout: options.timeout },

			logLevel: 'debug',
			viewportSize: {
				width: options.width,
				height: options.height } };





		if (_.has(options, 'loadImages')) {
			casperOptions.pageSettings.loadImages = options.loadImages;
		}

		this.__casper = casper.create(casperOptions);

		if (options.whitelist.length > 0 || options.blacklist.length > 0) {
			this.__casper.on('resource.requested', function (request, net) {
				if (options.whitelist.length > 0) {
					var found = false;var _iteratorNormalCompletion = true;var _didIteratorError = false;var _iteratorError = undefined;try {
						for (var _iterator = (0, _getIterator3.default)(options.whitelist), _step; !(_iteratorNormalCompletion = (_step = _iterator.next()).done); _iteratorNormalCompletion = true) {var white = _step.value;
							if (typeof white === 'string') {
								var url = request.url.toLowerCase();
								if (url.indexOf(white) === 0 || url.indexOf("https://" + white) === 0 || url.indexOf("http://" + white) === 0) {
									found = true;
									break;
								}
							} else if (white.test(request.url)) {
								found = true;
								break;
							}
						}} catch (err) {_didIteratorError = true;_iteratorError = err;} finally {try {if (!_iteratorNormalCompletion && _iterator.return) {_iterator.return();}} finally {if (_didIteratorError) {throw _iteratorError;}}}
					if (!found) {
						if (options.printAborts) {
							console.log("> Tab " + _this.id + ": Aborted (not found in whitelist): " + request.url);
						}
						return net.abort();
					}
				}var _iteratorNormalCompletion2 = true;var _didIteratorError2 = false;var _iteratorError2 = undefined;try {
					for (var _iterator2 = (0, _getIterator3.default)(options.blacklist), _step2; !(_iteratorNormalCompletion2 = (_step2 = _iterator2.next()).done); _iteratorNormalCompletion2 = true) {var black = _step2.value;
						if (typeof black === 'string') {
							var _url = request.url.toLowerCase();
							if (_url.indexOf(black) === 0 || _url.indexOf("https://" + black) === 0 || _url.indexOf("http://" + black) === 0) {
								if (options.printAborts) {
									console.log("> Tab " + _this.id + ": Aborted (blacklisted by \"" + black + "\"): " + _url);
								}
								return net.abort();
							}
						} else if (black.test(request.url)) {
							if (options.printAborts) {
								console.log("> Tab " + _this.id + ": Aborted (blacklisted by " + black + "): " + request.url);
							}
							return net.abort();
						}
					}} catch (err) {_didIteratorError2 = true;_iteratorError2 = err;} finally {try {if (!_iteratorNormalCompletion2 && _iterator2.return) {_iterator2.return();}} finally {if (_didIteratorError2) {throw _iteratorError2;}}}
			});
		}

		if (options.printNavigation) {
			this.__casper.on('navigation.requested', function (url, type, isLocked, isMainFrame) {
				if (isMainFrame) {
					console.log("> Tab " + _this.id + ": Navigation" + (type !== 'Other' ? " (" + type + ")" : '') + (isLocked ? '' : ' (not locked)') + ": " + url);
				}
			});
		}

		this.__casper.on('page.created', function (page) {
			if (options.httpProxy) {
				page.setProxy(options.httpProxy);
				console.log("CasperJS page proxy set to " + options.httpProxy);
			}
			if (options.printNavigation) {
				console.log("> Tab " + _this.id + ": New PhantomJS WebPage created");
				page.onResourceTimeout = function (request) {return console.log("> Tab " + _this.id + ": Timeout: " + request.url);};
			}
		});

		if (options.printPageErrors) {
			this.__casper.on('page.error', function (err) {
				console.log("> Tab " + _this.id + ": Page JavaScript error: " + err);
			});
		}

		if (options.printResourceErrors) {
			this.__casper.on('resource.error', function (err) {
				if (err.errorString === 'Protocol "" is unknown') {
					return;
				}
				var message = "> Tab " + _this.id + ": Resource error: " + (err.status != null ? err.status + " - " : '') + (err.statusText != null ? err.statusText + " - " : '') + err.errorString;
				if (typeof err.url === 'string' && message.indexOf(err.url) < 0) {
					message += " (" + err.url + ")";
				}
				console.log(message);
			});
		}


		this.__last50Errors = [];
		this.__openInProgress = false;
		this.__injectInProgress = false;
		this.__fetchState = {
			error: null,
			httpCode: null,
			httpStatus: null,
			url: null };



		this.__casper.on('resource.error', function (error) {
			if (_this.__openInProgress || _this.__injectInProgress) {
				_this.__last50Errors.push(error);
				if (_this.__last50Errors.length > 50) {
					_this.__last50Errors.shift();
				}
			}
		});



		this.__casper.on('page.resource.received', function (resource) {
			if (_this.__openInProgress) {
				if (typeof resource.status !== 'number') {
					_this.__fetchState.error = 'unknown error';
					if (typeof resource.id === 'number') {var _iteratorNormalCompletion3 = true;var _didIteratorError3 = false;var _iteratorError3 = undefined;try {
							for (var _iterator3 = (0, _getIterator3.default)(_this.__last50Errors), _step3; !(_iteratorNormalCompletion3 = (_step3 = _iterator3.next()).done); _iteratorNormalCompletion3 = true) {var err = _step3.value;
								if (resource.id === err.id) {
									if (typeof err.errorString === 'string') {
										_this.__fetchState.error = err.errorString;
									}
								}
							}} catch (err) {_didIteratorError3 = true;_iteratorError3 = err;} finally {try {if (!_iteratorNormalCompletion3 && _iterator3.return) {_iterator3.return();}} finally {if (_didIteratorError3) {throw _iteratorError3;}}}
					}
				} else {
					_this.__fetchState.httpCode = resource.status;
				}
				_this.__fetchState.httpStatus = resource.statusText;
				_this.__fetchState.url = resource.url;
			}
		});




		this.__casper.on('resource.received', function (resource) {
			if (_this.__injectInProgress) {
				if (resource.url === _this.__fetchState.url) {
					if (typeof resource.redirectURL === 'string') {
						_this.__fetchState.url = resource.redirectURL;
					} else if (resource.stage === 'end') {
						_this.__fetchState.httpCode = resource.status;
						_this.__fetchState.httpStatus = resource.statusText;
						_this.__fetchState.url = resource.url;
						if (typeof resource.status !== 'number') {
							_this.__fetchState.error = 'unknown error';
							if (typeof resource.id === 'number') {var _iteratorNormalCompletion4 = true;var _didIteratorError4 = false;var _iteratorError4 = undefined;try {
									for (var _iterator4 = (0, _getIterator3.default)(_this.__last50Errors), _step4; !(_iteratorNormalCompletion4 = (_step4 = _iterator4.next()).done); _iteratorNormalCompletion4 = true) {var err = _step4.value;
										if (resource.id === err.id) {
											if (typeof err.errorString === 'string') {
												_this.__fetchState.error = err.errorString;
											}
										}
									}} catch (err) {_didIteratorError4 = true;_iteratorError4 = err;} finally {try {if (!_iteratorNormalCompletion4 && _iterator4.return) {_iterator4.return();}} finally {if (_didIteratorError4) {throw _iteratorError4;}}}
							}
						} else if (resource.status < 200 || resource.status >= 300) {
							_this.__fetchState.error = "got HTTP " + resource.status + " " + resource.statusText + " when downloading " + resource.url;
						}
						_this.__injectInProgress = false;
						_this.__last50Errors = [];
					}
				}
			}
		});




		phantom.onError = function (msg, trace) {
			console.log("\n" + msg);
			if (trace && trace.length) {var _iteratorNormalCompletion5 = true;var _didIteratorError5 = false;var _iteratorError5 = undefined;try {
					for (var _iterator5 = (0, _getIterator3.default)(trace), _step5; !(_iteratorNormalCompletion5 = (_step5 = _iterator5.next()).done); _iteratorNormalCompletion5 = true) {var _f = _step5.value;
						console.log(" at " + (_f.file || _f.sourceURL) + ":" + _f.line + (_f.function ? " (in function " + _f.function + ")" : ''));
					}} catch (err) {_didIteratorError5 = true;_iteratorError5 = err;} finally {try {if (!_iteratorNormalCompletion5 && _iterator5.return) {_iterator5.return();}} finally {if (_didIteratorError5) {throw _iteratorError5;}}}
			}
			console.log('');
			phantom.exit(1);
		};




		this.__casper.start(null, null);
		var waitLoop = function waitLoop() {
			_this.__casper.wait(10);
			_this.__casper.then(function () {
				if (_this.__endCallback == null) {
					if (_this.__nextStep != null) {
						var step = _this.__nextStep;
						_this.__nextStep = null;
						step();
					}
					waitLoop();
				}
			});
		};
		waitLoop();
		this.__casper.run(function () {

			_this.__closed = true;
			if (typeof _this.__endCallback === 'function') {
				_this.__endCallback(null);
			}


			_this.__endCallback = null;
			_this.__casper.removeAllListeners('resource.requested');
			_this.__casper.removeAllListeners('navigation.requested');
			_this.__casper.removeAllListeners('page.created');
			_this.__casper.removeAllListeners('page.error');
			_this.__casper.removeAllListeners('resource.error');
			_this.__casper.removeAllListeners('page.resource.received');
			_this.__casper.removeAllListeners('resource.received');
			_this.__casper.page.clearMemoryCache();
			_this.__casper.page.close();
			delete _this.__casper.page;
			_this.__casper = null;
		});


		this.__casper.setFilter("page.confirm", function (msg) {
			return _this._onConfirm(msg);
		});
		this.__casper.setFilter("page.prompt", function (msg) {
			return _this._onPrompt(msg);
		});
	}(0, _createClass3.default)(TabDriver, [{ key: "_close", value: function _close(








		callback) {
			this.__endCallback = callback;
		} }, { key: "_open", value: function _open(

		url, options, callback) {var _this2 = this;
			this.__nextStep = function () {
				_this2.__casper.clear();
				_this2.__openInProgress = true;
				_this2.__fetchState.error = null;
				_this2.__fetchState.httpCode = null;
				_this2.__fetchState.httpStatus = null;
				_this2.__fetchState.url = null;
				_this2.__casper.thenOpen(url, options);
				_this2.__casper.then(function () {
					_this2.__openInProgress = false;
					_this2.__last50Errors = [];


					if (_this2.__fetchState.error != null || _this2.__fetchState.httpCode != null) {
						callback(_this2.__fetchState.error, _this2.__fetchState.httpCode, _this2.__fetchState.httpStatus, _this2.__fetchState.url);
					} else {
						if (url.trim().toLowerCase().indexOf('file://') === 0) {

							callback(null, null, _this2.__fetchState.httpStatus, _this2.__fetchState.url);
						} else {
							callback('unknown error', null, _this2.__fetchState.httpStatus, _this2.__fetchState.url);
						}
					}
				});
			};
		} }, { key: "_waitUntilVisible", value: function _waitUntilVisible(

		selectors, duration, condition, callback) {this.__callCasperWaitMethod('waitUntilVisible', selectors, duration, condition, callback);} }, { key: "_waitWhileVisible", value: function _waitWhileVisible(
		selectors, duration, condition, callback) {this.__callCasperWaitMethod('waitWhileVisible', selectors, duration, condition, callback);} }, { key: "_waitUntilPresent", value: function _waitUntilPresent(
		selectors, duration, condition, callback) {this.__callCasperWaitMethod('waitForSelector', selectors, duration, condition, callback);} }, { key: "_waitWhilePresent", value: function _waitWhilePresent(
		selectors, duration, condition, callback) {this.__callCasperWaitMethod('waitWhileSelector', selectors, duration, condition, callback);} }, { key: "__callCasperWaitMethod", value: function __callCasperWaitMethod(
		method, selectors, duration, condition, callback) {var _this3 = this;
			this.__nextStep = function () {
				var start = Date.now();
				var index = 0;
				if (condition === 'and') {
					var nextSelector = function nextSelector() {
						var success = function success() {
							++index;
							if (index >= selectors.length) {
								callback(null, null);
							} else {
								duration -= Date.now() - start;
								if (duration < _this3.__casper.options.retryTimeout * 2) {
									duration = _this3.__casper.options.retryTimeout * 2;
								}
								nextSelector();
							}
						};
						var failure = function failure() {
							callback("waited " + (Date.now() - start) + "ms but element \"" + selectors[index] + "\" still " + (method.indexOf('While') > 0 ? '' : 'not ') + (method.indexOf('Visible') > 0 ? 'visible' : 'present'));
						};
						_this3.__casper[method](selectors[index], success, failure, duration);
					};
				} else {
					var waitedForAll = false;
					var nextSelector = function nextSelector() {
						var success = function success() {
							callback(null, selectors[index]);
						};
						var failure = function failure() {
							if (waitedForAll && start + duration < Date.now()) {
								var elementsToString = selectors.slice();var _iteratorNormalCompletion6 = true;var _didIteratorError6 = false;var _iteratorError6 = undefined;try {
									for (var _iterator6 = (0, _getIterator3.default)(elementsToString), _step6; !(_iteratorNormalCompletion6 = (_step6 = _iterator6.next()).done); _iteratorNormalCompletion6 = true) {var e = _step6.value;
										e = "\"" + e + "\"";
									}} catch (err) {_didIteratorError6 = true;_iteratorError6 = err;} finally {try {if (!_iteratorNormalCompletion6 && _iterator6.return) {_iterator6.return();}} finally {if (_didIteratorError6) {throw _iteratorError6;}}}
								elementsToString = elementsToString.join(', ');
								callback("waited " + (Date.now() - start) + "ms but element" + (selectors.length > 0 ? 's' : '') + " " + elementsToString + " still " + (method.indexOf('While') > 0 ? '' : 'not ') + (method.indexOf('Visible') > 0 ? 'visible' : 'present'), null);
							} else {
								++index;
								if (index >= selectors.length) {
									waitedForAll = true;
									index = 0;
								}
								nextSelector();
							}
						};
						_this3.__casper[method](selectors[index], success, failure, _this3.__casper.options.retryTimeout * 2);
					};
				}
				nextSelector();
			};
		} }, { key: "_click", value: function _click(

		selector, options, callback) {var _this4 = this;
			this.__nextStep = function () {
				try {

					_this4.__casper.click(selector);
					callback(null);
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "_evaluate", value: function _evaluate(

		func, arg, callback) {var _this5 = this;
			this.__nextStep = function () {
				var err = null;
				try {
					f = function f(__arg, __code) {
						var cb = function cb(err, res) {
							window.__evaluateAsyncFinished = true;
							window.__evaluateAsyncErr = err;
							window.__evaluateAsyncRes = res;
						};
						try {
							window.__evaluateAsyncFinished = false;
							window.__evaluateAsyncErr = null;
							window.__evaluateAsyncRes = null;
							eval("(" + __code + ")")(__arg, cb);
							return undefined;
						} catch (e) {
							return e.toString();
						}
					};
					err = _this5.__casper.evaluate(f, arg, func.toString());
					if (err != null) {
						err = "in evaluated code (initial call): " + err;
					}
				} catch (e) {
					err = "in casper context (initial call): " + e.toString();
				}
				if (err != null) {
					callback(err, null);
				} else {
					check = function (_check) {function check() {return _check.apply(this, arguments);}check.toString = function () {return _check.toString();};return check;}(function () {
						try {
							var res = _this5.__casper.evaluate(function () {

								return {
									finished: window.__evaluateAsyncFinished,
									err: window.__evaluateAsyncErr != null ? window.__evaluateAsyncErr : undefined,
									res: window.__evaluateAsyncRes != null ? window.__evaluateAsyncRes : undefined };

							});
							if (res.finished) {
								callback(res.err === undefined ? null : res.err, res.res === undefined ? null : res.res);
							} else {
								setTimeout(check, 200);
							}
						} catch (e) {
							callback("in casper context (callback): " + e.toString(), null);
						}
					});
					setTimeout(check, 100);
				}
			};
		} }, { key: "_getUrl", value: function _getUrl(

		callback) {var _this6 = this;
			this.__nextStep = function () {
				try {
					callback(null, _this6.__casper.getCurrentUrl());
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "_getContent", value: function _getContent(

		callback) {var _this7 = this;
			this.__nextStep = function () {
				try {
					callback(null, _this7.__casper.getPageContent());
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "_fill", value: function _fill(

		selector, params, options, callback) {var _this8 = this;
			this.__nextStep = function () {
				try {
					_this8.__casper.fill(selector, params, options.submit);
					callback(null);
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "_screenshot", value: function _screenshot(

		filename, options, callback) {var _this9 = this;
			this.__nextStep = function () {
				try {



					_this9.__casper.capture(filename);
					callback(null, filename);
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "_sendKeys", value: function _sendKeys(

		selector, keys, options, callback) {var _this10 = this;
			this.__nextStep = function () {
				try {
					_this10.__casper.sendKeys(selector, keys, options);
					callback(null);
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "_injectFromDisk", value: function _injectFromDisk(

		path, callback) {var _this11 = this;
			this.__nextStep = function () {
				try {


					var ret = _this11.__casper.page.injectJs(path);
				} catch (e) {
					callback("failed to inject local script \"" + path + "\": " + e.toString());
					return;
				}
				if (ret) {
					callback(null);
				} else {
					callback("failed to inject local script \"" + path + "\"");
				}
			};
		} }, { key: "_injectFromUrl", value: function _injectFromUrl(

		url, callback) {var _this12 = this;
			this.__nextStep = function () {
				_this12.__injectInProgress = true;
				_this12.__fetchState.url = url.trim();
				try {


					_this12.__casper.page.includeJs(_this12.__fetchState.url);
				} catch (e) {
					_this12.__injectInProgress = false;
					callback(e.toString());
					return;
				}
				_this12.__fetchState.error = null;
				_this12.__fetchState.httpCode = null;
				_this12.__fetchState.httpStatus = null;
				var injectStart = Date.now();
				waitForInject = function (_waitForInject) {function waitForInject() {return _waitForInject.apply(this, arguments);}waitForInject.toString = function () {return _waitForInject.toString();};return waitForInject;}(function () {
					setTimeout(function () {
						if (_this12.__injectInProgress) {

							if (Date.now() - injectStart > _this12.__casper.page.settings.resourceTimeout + 1000) {
								callback("injection of script \"" + _this12.__fetchState.url + "\" timed out after " + (Date.now() - injectStart) + "ms");
							} else {
								waitForInject();
							}
						} else {

							callback(_this12.__fetchState.error, _this12.__fetchState.httpCode, _this12.__fetchState.httpStatus, _this12.__fetchState.url);
						}
					}, 100);
				});
				waitForInject();
			};
		} }, { key: "_scroll", value: function _scroll(

		x, y, callback) {var _this13 = this;
			this.__nextStep = function () {
				try {
					_this13.__casper.scrollTo(x, y);
					callback(null);
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "_scrollToBottom", value: function _scrollToBottom(

		callback) {var _this14 = this;
			this.__nextStep = function () {
				try {
					_this14.__casper.scrollToBottom();
					callback(null);
				} catch (e) {
					callback(e.toString());
				}
			};
		} }, { key: "casper", get: function get() {return this.__casper;} }, { key: "closed", get: function get() {return this.__closed;} }, { key: "crashed", get: function get() {return this.__crashed;} }, { key: "id", get: function get() {return this.__uniqueTabId;} }]);return TabDriver;}();



module.exports = TabDriver;
